presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java

```diff
@@ -725,9 +729,51 @@ private FunctionCall rewriteCountAsSum(FunctionCall node, Expression derivedColu
                     ImmutableList.of(derivedColumnName));
         }
 
+        private Expression rewriteExpressionWithCoercions(Expression expression, ExpressionAnalysis analysis)
+        {
+            return ExpressionTreeRewriter.rewriteWith(new ExpressionRewriter<Void>()
+            {
+                @Override
+                public Expression rewriteExpression(Expression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+                {
+                    Expression rewrittenExpression = treeRewriter.defaultRewrite(node, context);
+                    return coerceIfNecessary(node, rewrittenExpression);
+                }
+
+                private Expression coerceIfNecessary(Expression original, Expression rewritten)
+                {
+                    Type coercion = analysis.getCoercion(original);
+                    if (coercion != null) {
+                        rewritten = new Cast(
+                                original.getLocation(),
+                                rewritten,
+                                coercion.getTypeSignature().toString(),
+                                false,
+                                analysis.isTypeOnlyCoercion(original));
+                    }
+                    return rewritten;
+                }
+            }, expression, null);
```

Comment:
This looks odd, create a separate class within this class which extends ExpressionRewriter